sw/cheri/common/platform-pinmux.hh.tpl

```diff
@@ -1,195 +1,201 @@
-/**
- * Copyright lowRISC contributors.
- * Licensed under the Apache License, Version 2.0, see LICENSE for details.
- * SPDX-License-Identifier: Apache-2.0
- */
+// SPDX-FileCopyrightText: lowRISC contributors
+// SPDX-License-Identifier: Apache-2.0
 
 ${"#"}pragma once
 ${"#"}include <debug.hh>
 ${"#"}include <stdint.h>
 ${"#"}include <utils.hh>
+${"#"}include <cheri.hh>
 
-/**
- * A driver for Sonata's Pin Multiplexer (Pinmux).
- *
- * This driver can be used to select which block output is placed on a given
- * output pin, and to select which input pin is provided to a given block
- * input. Sonata's pinmux only allows certain selections per output pin / block
- * input, which this driver describes.
- *
- * Rendered documentation is served from:
- * https://lowrisc.github.io/sonata-system/doc/ip/pinmux.html
- */
-class SonataPinmux : private utils::NoCopyNoMove {
-  /**
-   * Flag to set when debugging the driver for UART log messages.
-   */
-  static constexpr bool DebugDriver = false;
+namespace SonataPinmux {
+/// The number of pin sinks (pin outputs)
+static constexpr size_t NumPinSinks = ${len(output_pins)};
 
-  /**
-   * Helper for conditional debug logs and assertions.
-   */
-  using Debug = ConditionalDebug<DebugDriver, "Pinmux">;
+/// The number of block sinks (block inputs)
+static constexpr size_t NumBlockSinks = ${len(output_block_ios)};
 
-  /**
-   * A pointer/capability to the Pin Multiplexer's registers, where
-   * each sequential byte potentially corresponds to some mapped pinmux
-   * selector used to select the pin configuration.
-   */
-  volatile uint8_t *registers;
+/// Flag to set when debugging the driver for UART log messages.
+static constexpr bool DebugDriver = false;
 
- public:
-  /**
-   * The Output Pins defined for the Sonata board. These output pins can be
-   * multiplexed, meaning that they can be changed to output the outputs
-   * of different blocks (or disabled). The block outputs that can be
-   * selected are limited, and vary on a per-pin basis.
-   *
-   * Documentation sources:
-   * https://lowrisc.github.io/sonata-system/doc/ip/pinmux.html
-   * https://github.com/lowRISC/sonata-system/blob/4b72d8c07c727846c6ccb27754352388f3b2ac9a/data/pins_sonata.xdc
-   * https://github.com/newaetech/sonata-pcb/blob/649b11c2fb758f798966605a07a8b6b68dd434e9/sonata-schematics-r09.pdf
-   */
-  enum class OutputPin : uint16_t {
+/// Helper for conditional debug logs and assertions.
+using Debug = ConditionalDebug<DebugDriver, "Pinmux">;
+
+/**
+ * Each pin sink is configured by an 8-bit register. This enum maps pin sink
+ * names to the offset of their configuration registers. The offsets are relative
+ * to the first pin sink register.
+ *
+ * Documentation sources:
+ * 1. https://lowrisc.github.io/sonata-system/doc/ip/pinmux/
+ * 2. https://github.com/lowRISC/sonata-system/blob/4b72d8c07c727846c6ccb27754352388f3b2ac9a/data/pins_sonata.xdc
```

Comment:
Might be nice to bump this permalink as it is a bit out of date, and doesn't include e.g. the PMODC or RS485 pins and a couple of other changes.